File: cli-exec.js
/* QPT Workbench — command language over the store.
 *
 * One executor shared by the HTTP endpoint (POST /api/cli), the standalone bin
 * (node cli.js "..."), and available to the agent. Mutations run through
 * qpt-core.js — the same enforcement as every other surface. Skills live on
 * disk in the Agent Skills format (data/skills/<name>/SKILL.md).
 *
 *   help · state · boards · board <id> · cards [--board b] · card <id>
 *   create --board b --title "…" [--source s] [--target t] [--kind k] [--reliability n] [--tags a,b] [--note "…"]
 *   move <id> <column> · edit <id> key=value… · evaluate <id> · promote <id> --to <board> · reset
 *   skills · skill show <id> · skill create <id> --description "…" [--content "…"]
 *   skill edit <id> [name=…] [description=…] [content=…] · skill delete <id>
 */

import { listSkills, readSkill, createSkill, updateSkill, deleteSkill } from "./server-skills.js";

export function tokenize(s) {
  const out = [];
  // key="quoted value" | key='quoted value' | key=value | "quoted" | 'quoted' | bare
  const re = /(\S+?=)(?:"([^"]*)"|'([^']*)'|(\S+))|"([^"]*)"|'([^']*)'|(\S+)/g;
  let m;
  while ((m = re.exec(s))) {
    if (m[1] !== undefined) out.push(m[1] + (m[2] ?? m[3] ?? m[4] ?? ""));
    else out.push(m[5] ?? m[6] ?? m[7]);
  }
  return out;
}

const HELP = `qpt workbench CLI
  state                              compact store snapshot (JSON)
  boards                             boards and their column flows
  board <id>                         switch the active board (protocol|dialectic|resolution)
  cards [--board <id>]               list cards on a board
  card <id>                          full card JSON
  create --board <b> --title "…"     new card at the board's entry column
       [--source recruited|initiated] [--target grounded|terminated]
       [--kind position|note] [--reliability 0..1] [--tags a,b] [--note "…"]
  move <id> <column>                 move one column (A13: no forward skips)
  edit <id> key=value …              domain-clamped edit (rho delta gamma k theta
                                     source target scale pathology death reliability genesis title note sign agent tags)
  evaluate <id>                      gate verdict with S vs θ
  promote <id> --to <board>          cross-board promotion (paths per §15/§7.4/§11/§29)
  reset                              reseed the store from the specification
  skills                             list skills (data/skills/<name>/SKILL.md)
  skill show <id>                    print a skill
  skill create <id> --description "…" [--content "…"]
  skill edit <id> [name=…] [description=…] [content=…]
  skill delete <id>`;

function findSkill(skills, idOrName) {
  return skills.find((s) => s.id === idOrName) || skills.find((s) => s.name === idOrName);
}

export async function execCommand(line, store) {
  const t = tokenize(line);
  const ok = (output, changed = false) => ({ ok: true, output, changed });
  const bad = (output) => ({ ok: false, output, changed: false });
  if (!t.length) return bad("empty command — try `help`");

  const [cmd, ...rest] = t;
  const pos = [];
  const flags = {};
  for (let i = 0; i < rest.length; i++) {
    if (rest[i].startsWith("--")) {
      const k = rest[i].slice(2);
      flags[k] = rest[i + 1] != null && !rest[i + 1].startsWith("--") ? rest[++i] : true;
    } else pos.push(rest[i]);
  }

  const CORE = store.CORE;
  const data = store.data;

  async function finish(r) {
    if (r.ok) { await store.save(); return ok(r.message, true); }
    return bad(r.message);
  }

  switch (cmd) {
    case "help":
      return ok(HELP);

    case "state":
      return ok(JSON.stringify(CORE.compactState(data), null, 1));

    case "boards":
      return ok(CORE.boards.map((b) => b.id + ":  " + b.columns.map((c) => c.id).join(" → ")).join("\n"));

    case "board": {
      if (!CORE.boards.some((b) => b.id === pos[0])) return bad("unknown board " + (pos[0] || "(missing)"));
      data.boardId = pos[0];
      await store.save();
      return ok("board → " + pos[0], true);
    }

    case "cards": {
      const boardId = flags.board || data.boardId;
      const list = CORE.cardsOn(data, boardId);
      if (!list.length) return ok("(no cards on " + boardId + ")");
      return ok(list.map((c) => {
        const ev = CORE.evaluate(c);
        return c.id + " @ " + c.column + " — " + c.title + (ev ? "  [" + ev.key + "]" : "");
      }).join("\n"));
    }

    case "card": {
      const c = data.cards[pos[0]];
      if (!c) return bad("no card " + (pos[0] || "(missing id)"));
      return ok(JSON.stringify(c, null, 1));
    }

    case "create":
      return finish(CORE.applyAction(data, {
        action: "create_card",
        board: flags.board || data.boardId,
        title: flags.title,
        source: flags.source,
        target: flags.target,
        kind: flags.kind,
        reliability: flags.reliability != null ? parseFloat(flags.reliability) : undefined,
        tags: flags.tags ? String(flags.tags).split(",").map((x) => x.trim()).filter(Boolean) : undefined,
        note: flags.note,
        via: "the CLI",
      }));

    case "move":
      if (!pos[0] || !pos[1]) return bad("usage: move <id> <column>");
      return finish(CORE.applyAction(data, { action: "move_card", id: pos[0], column: pos[1] }));

    case "edit": {
      if (!pos[0]) return bad("usage: edit <id> key=value …");
      const patch = {};
      for (const kv of pos.slice(1)) {
        const eq = kv.indexOf("=");
        if (eq < 0) return bad("edit: expected key=value, got " + JSON.stringify(kv));
        let v = kv.slice(eq + 1);
        if (v === "null") v = null;
        else if (v === "true") v = true;
        else if (v === "false") v = false;
        else if (v !== "" && !isNaN(+v)) v = +v;
        else if (v.indexOf(",") >= 0) v = v.split(",").map((x) => x.trim()).filter(Boolean);
        patch[kv.slice(0, eq)] = v;
      }
      return finish(CORE.applyAction(data, { action: "edit_card", id: pos[0], patch }));
    }

    case "evaluate": {
      const r = CORE.applyAction(data, { action: "evaluate_card", id: pos[0] });
      return r.ok ? ok(r.message) : bad(r.message);
    }

    case "promote":
      if (!pos[0] || !flags.to) return bad("usage: promote <id> --to <board>");
      return finish(CORE.applyAction(data, { action: "promote_card", id: pos[0], board: flags.to }));

    case "reset":
      await store.reset();
      return ok("store reseeded from the specification", true);

    case "skills": {
      const { skills, warnings } = await listSkills();
      const lines = skills.map((s) => s.id + " — " + s.description.slice(0, 80) + (s.description.length > 80 ? "…" : ""));
      warnings.forEach((w) => lines.push("⚠ " + w));
      return ok(lines.length ? lines.join("\n") : "(no skills — data/skills/ is empty)");
    }

    case "skill": {
      const sub = pos[0];
      try {
        if (sub === "show") {
          const { skills } = await listSkills();
          const hit = findSkill(skills, pos[1]);
          if (!hit) return bad("no skill " + (pos[1] || "(missing id)"));
          const s = await readSkill(hit.id);
          return ok("# " + s.name + " (" + s.id + ")\n" + s.description + "\n\n" + s.content);
        }
        if (sub === "create") {
          if (!pos[1]) return bad("usage: skill create <id> --description \"…\" [--content \"…\"]");
          const s = await createSkill({
            name: pos[1],
            description: flags.description || "",
            content: flags.content || "",
          });
          return ok("skill created: data/skills/" + s.id + "/SKILL.md", true);
        }
        if (sub === "edit") {
          const { skills } = await listSkills();
          const hit = findSkill(skills, pos[1]);
          if (!hit) return bad("no skill " + (pos[1] || "(missing id)"));
          const patch = {};
          const allowed = ["name", "description", "content"];
          for (const kv of pos.slice(2)) {
            const eq = kv.indexOf("=");
            if (eq < 0) return bad("skill edit: expected key=value, got " + JSON.stringify(kv));
            const k = kv.slice(0, eq);
            if (!allowed.includes(k)) return bad("skill edit: key must be " + allowed.join("|"));
            patch[k] = kv.slice(eq + 1);
          }
          if (!Object.keys(patch).length) return bad("skill edit: nothing to change");
          const s = await updateSkill(hit.id, patch);
          return ok("skill updated: " + s.id, true);
        }
        if (sub === "delete") {
          const { skills } = await listSkills();
          const hit = findSkill(skills, pos[1]);
          if (!hit) return bad("no skill " + (pos[1] || "(missing id)"));
          await deleteSkill(hit.id);
          return ok("skill deleted: " + hit.id, true);
        }
        return bad("skill: expected show|create|edit|delete");
      } catch (e) {
        return bad(String((e && e.message) || e));
      }
    }

    default:
      return bad("unknown command " + JSON.stringify(cmd) + " — try `help`");
  }
}

File: rust_ui/src/cli_exec.rs
//! Port of cli-exec.js — the command language over the store.
//!
//! One executor shared by the HTTP endpoint (POST /api/cli), the standalone
//! bin (qpt-cli "..."), and available to the agent. Mutations run through
//! `core` — the same enforcement as every other surface.

use serde_json::{json, Map, Value};

use crate::{core, skills, store::Store, Paths};

pub struct CliResult {
    pub ok: bool,
    pub output: String,
    pub changed: bool,
}

fn ok(output: String, changed: bool) -> CliResult {
    CliResult { ok: true, output, changed }
}

fn bad(output: String) -> CliResult {
    CliResult { ok: false, output, changed: false }
}

pub fn tokenize(s: &str) -> Vec<String> {
    // key="quoted value" | key='quoted value' | key=value | "quoted" | 'quoted' | bare
    static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
    let re = RE.get_or_init(|| regex::Regex::new(r#"(\S+?=)(?:"([^"]*)"|'([^']*)'|(\S+))|"([^"]*)"|'([^']*)'|(\S+)"#).unwrap());
    let mut out = Vec::new();
    for m in re.captures_iter(s) {
        if let Some(key) = m.get(1) {
            let val = m.get(2).or(m.get(3)).or(m.get(4)).map(|v| v.as_str()).unwrap_or("");
            out.push(format!("{}{}", key.as_str(), val));
        } else {
            let val = m.get(5).or(m.get(6)).or(m.get(7)).map(|v| v.as_str()).unwrap_or("");
            out.push(val.to_string());
        }
    }
    out
}

const HELP: &str = r#"qpt workbench CLI
  state                              compact store snapshot (JSON)
  boards                             boards and their column flows
  board <id>                         switch the active board (protocol|dialectic|resolution)
  cards [--board <id>]               list cards on a board
  card <id>                          full card JSON
  create --board <b> --title "…"     new card at the board's entry column
       [--source recruited|initiated] [--target grounded|terminated]
       [--kind position|note] [--reliability 0..1] [--tags a,b] [--note "…"]
  move <id> <column>                 move one column (A13: no forward skips)
  edit <id> key=value …              domain-clamped edit (rho delta gamma k theta
                                     source target scale pathology death reliability genesis title note sign agent tags)
  evaluate <id>                      gate verdict with S vs θ
  promote <id> --to <board>          cross-board promotion (paths per §15/§7.4/§11/§29)
  reset                              reseed the store from the specification
  skills                             list skills (data/skills/<name>/SKILL.md)
  skill show <id>                    print a skill
  skill create <id> --description "…" [--content "…"]
  skill edit <id> [name=…] [description=…] [content=…]
  skill delete <id>"#;

/// Flags: `--key value` or bare `--key` → true, as in the JS parser.
type Flags = Map<String, Value>;

fn flag_str<'a>(flags: &'a Flags, k: &str) -> Option<&'a str> {
    flags.get(k).and_then(Value::as_str)
}

/// JS `flags.board || data.boardId` — a bare `--board` is `true` in JS.
fn flag_or(flags: &Flags, k: &str, default: &str) -> String {
    match flags.get(k) {
        Some(Value::String(s)) => s.clone(),
        Some(Value::Bool(true)) => "true".into(),
        _ => default.to_string(),
    }
}

fn js_json_str(s: &str) -> String {
    serde_json::to_string(s).unwrap_or_default()
}

pub fn exec_command(line: &str, store: &mut Store, paths: &Paths) -> CliResult {
    let t = tokenize(line);
    if t.is_empty() {
        return bad("empty command — try `help`".into());
    }

    let cmd = t[0].clone();
    let rest = &t[1..];
    let mut pos: Vec<String> = Vec::new();
    let mut flags: Flags = Map::new();
    let mut i = 0;
    while i < rest.len() {
        if let Some(k) = rest[i].strip_prefix("--") {
            let next = rest.get(i + 1);
            match next {
                Some(n) if !n.starts_with("--") => {
                    flags.insert(k.to_string(), Value::String(n.clone()));
                    i += 2;
                }
                _ => {
                    flags.insert(k.to_string(), Value::Bool(true));
                    i += 1;
                }
            }
        } else {
            pos.push(rest[i].clone());
            i += 1;
        }
    }

    /// Run an action through the core; save on success.
    macro_rules! finish {
        ($action:expr) => {{
            let r = core::apply_action(&mut store.data, &$action);
            if r.get("ok") == Some(&json!(true)) {
                store.save();
                ok(crate::core::strf(&r, "message").unwrap_or("").to_string(), true)
            } else {
                bad(crate::core::strf(&r, "message").unwrap_or("").to_string())
            }
        }};
    }

    match cmd.as_str() {
        "help" => ok(HELP.to_string(), false),

        "state" => {
            ok(crate::json_indent1(&core::compact_state(&store.data)), false)
        }

        "boards" => ok(
            crate::data::boards()
                .iter()
                .map(|b| {
                    format!(
                        "{}:  {}",
                        b.id,
                        b.columns.iter().map(|c| c.id.clone()).collect::<Vec<_>>().join(" → ")
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"),
            false,
        ),

        "board" => {
            let id = pos.first().cloned().unwrap_or_default();
            if crate::data::boards().iter().all(|b| b.id != id) {
                return bad(format!(
                    "unknown board {}",
                    if id.is_empty() { "(missing)".to_string() } else { id }
                ));
            }
            store.data.board_id = id.clone();
            store.save();
            ok(format!("board → {}", id), true)
        }

        "cards" => {
            let board_id = flag_or(&flags, "board", &store.data.board_id);
            let list = core::cards_on(&store.data, &board_id);
            if list.is_empty() {
                return ok(format!("(no cards on {})", board_id), false);
            }
            ok(list
                .iter()
                .map(|c| {
                    let ev = core::evaluate(c);
                    format!(
                        "{}{} @ {} — {}{}",
                        core::strf(c, "id").unwrap_or(""),
                        core::strf(c, "handle")
                            .map(|h| format!(" @{}", h))
                            .unwrap_or_default(),
                        core::strf(c, "column").unwrap_or(""),
                        core::strf(c, "title").unwrap_or(""),
                        ev.map(|e| format!("  [{}]", e.key)).unwrap_or_default()
                    )
                })
                .collect::<Vec<_>>()
                .join("\n"), false)
        }

        "card" => {
            let reference = pos.first().cloned().unwrap_or_default();
            match core::find_card(&store.data.cards, &reference) {
                None => bad(format!(
                    "no card {}",
                    if reference.is_empty() { "(missing id)".to_string() } else { reference }
                )),
                Some(c) => ok(crate::json_indent1(c), false),
            }
        }

        "create" => {
            let mut action = Map::new();
            action.insert("action".into(), json!("create_card"));
            action.insert("board".into(), json!(flag_or(&flags, "board", &store.data.board_id)));
            for k in ["title", "source", "target", "kind", "note"] {
                if let Some(v) = flag_str(&flags, k) {
                    action.insert(k.into(), json!(v));
                }
            }
            if let Some(v) = flags.get("reliability") {
                let n = match v {
                    Value::String(s) => crate::js_parse_float(s),
                    Value::Bool(_) => None,
                    _ => v.as_f64(),
                };
                // JS parseFloat failure → NaN → serialized as null downstream
                action.insert("reliability".into(), n.map(|v| json!(v)).unwrap_or(Value::Null));
            }
            if let Some(tags) = flag_str(&flags, "tags") {
                let arr: Vec<Value> = tags
                    .split(',')
                    .map(|x| x.trim())
                    .filter(|x| !x.is_empty())
                    .map(|x| json!(x))
                    .collect();
                action.insert("tags".into(), Value::Array(arr));
            }
            action.insert("via".into(), json!("the CLI"));
            finish!(Value::Object(action))
        }

        "move" => {
            if pos.len() < 2 {
                return bad("usage: move <id|@handle> <column>".into());
            }
            let Some(c) = core::find_card(&store.data.cards, &pos[0]) else {
                return bad(format!("no card {}", pos[0]));
            };
            let id = core::strf(c, "id").unwrap_or("").to_string();
            finish!(json!({ "action": "move_card", "id": id, "column": pos[1] }))
        }

        "edit" => {
            if pos.is_empty() {
                return bad("usage: edit <id|@handle> key=value …".into());
            }
            let Some(c) = core::find_card(&store.data.cards, &pos[0]) else {
                return bad(format!("no card {}", pos[0]));
            };
            let id = core::strf(c, "id").unwrap_or("").to_string();
            let mut patch = Map::new();
            static NUM_RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
            let num_re = NUM_RE.get_or_init(|| regex::Regex::new(r"^[+-]?(\d+\.?\d*|\.\d+)([eE][+-]?\d+)?$").unwrap());
            for kv in &pos[1..] {
                let Some(eq) = kv.find('=') else {
                    return bad(format!("edit: expected key=value, got {}", js_json_str(kv)));
                };
                let v = &kv[eq + 1..];
                let value = if v == "null" {
                    Value::Null
                } else if v == "true" {
                    json!(true)
                } else if v == "false" {
                    json!(false)
                } else if !v.is_empty() && num_re.is_match(v) {
                    json!(v.parse::<f64>().unwrap_or(0.0))
                } else if v.contains(',') {
                    Value::Array(
                        v.split(',')
                            .map(|x| x.trim())
                            .filter(|x| !x.is_empty())
                            .map(|x| json!(x))
                            .collect(),
                    )
                } else {
                    json!(v)
                };
                patch.insert(kv[..eq].to_string(), value);
            }
            finish!(json!({ "action": "edit_card", "id": id, "patch": Value::Object(patch) }))
        }

        "evaluate" => {
            let reference = pos.first().cloned().unwrap_or_default();
            let Some(c) = core::find_card(&store.data.cards, &reference) else {
                return bad(format!(
                    "no card {}",
                    if reference.is_empty() { "(missing id)".to_string() } else { reference }
                ));
            };
            let id = core::strf(c, "id").unwrap_or("").to_string();
            let r = core::apply_action(&mut store.data, &json!({ "action": "evaluate_card", "id": id }));
            let msg = core::strf(&r, "message").unwrap_or("").to_string();
            if r.get("ok") == Some(&json!(true)) {
                ok(msg, false)
            } else {
                bad(msg)
            }
        }

        "promote" => {
            if pos.is_empty() || flags.get("to").is_none() {
                return bad("usage: promote <id|@handle> --to <board>".into());
            }
            let Some(c) = core::find_card(&store.data.cards, &pos[0]) else {
                return bad(format!("no card {}", pos[0]));
            };
            let id = core::strf(c, "id").unwrap_or("").to_string();
            let to = flag_or(&flags, "to", "");
            finish!(json!({ "action": "promote_card", "id": id, "board": to }))
        }

        "reset" => {
            store.reset();
            ok("store reseeded from the specification".into(), true)
        }

        "skills" => {
            let (skills_list, warnings) = skills::list_skills(paths);
            let mut lines: Vec<String> = skills_list
                .iter()
                .map(|s| {
                    let d = &s.description;
                    if d.chars().count() > 80 {
                        format!("{} — {}…", s.id, d.chars().take(80).collect::<String>())
                    } else {
                        format!("{} — {}", s.id, d)
                    }
                })
                .collect();
            for w in warnings {
                lines.push(format!("⚠ {}", w));
            }
            if lines.is_empty() {
                ok("(no skills — data/skills/ is empty)".into(), false)
            } else {
                ok(lines.join("\n"), false)
            }
        }

        "skill" => {
            let sub = pos.first().map(String::as_str).unwrap_or("");
            let run = || -> Result<CliResult, String> {
                match sub {
                    "show" => {
                        let (skills_list, _) = skills::list_skills(paths);
                        let hit = pos.get(1).and_then(|q| {
                            skills_list.iter().find(|s| &s.id == q || &s.name == q)
                        });
                        let Some(hit) = hit else {
                            return Ok(bad(format!(
                                "no skill {}",
                                pos.get(1).cloned().unwrap_or_else(|| "(missing id)".into())
                            )));
                        };
                        let s = skills::read_skill(paths, &hit.id)?;
                        Ok(ok(format!("# {} ({})\n{}\n\n{}", s.name, s.id, s.description, s.content), false))
                    }
                    "create" => {
                        if pos.get(1).is_none() {
                            return Ok(bad(
                                "usage: skill create <id> --description \"…\" [--content \"…\"]".into(),
                            ));
                        }
                        let s = skills::create_skill(
                            paths,
                            &json!({
                                "name": pos[1],
                                "description": flag_str(&flags, "description").unwrap_or(""),
                                "content": flag_str(&flags, "content").unwrap_or(""),
                            }),
                        )?;
                        Ok(ok(format!("skill created: data/skills/{}/SKILL.md", s.id), true))
                    }
                    "edit" => {
                        let (skills_list, _) = skills::list_skills(paths);
                        let hit = pos.get(1).and_then(|q| {
                            skills_list.iter().find(|s| &s.id == q || &s.name == q)
                        });
                        let Some(hit) = hit else {
                            return Ok(bad(format!(
                                "no skill {}",
                                pos.get(1).cloned().unwrap_or_else(|| "(missing id)".into())
                            )));
                        };
                        let mut patch = Map::new();
                        let allowed = ["name", "description", "content"];
                        for kv in &pos[2..] {
                            let Some(eq) = kv.find('=') else {
                                return Ok(bad(format!(
                                    "skill edit: expected key=value, got {}",
                                    js_json_str(kv)
                                )));
                            };
                            let k = &kv[..eq];
                            if !allowed.contains(&k) {
                                return Ok(bad(format!(
                                    "skill edit: key must be {}",
                                    allowed.join("|")
                                )));
                            }
                            patch.insert(k.to_string(), json!(kv[eq + 1..]));
                        }
                        if patch.is_empty() {
                            return Ok(bad("skill edit: nothing to change".into()));
                        }
                        let s = skills::update_skill(paths, &hit.id, &Value::Object(patch))?;
                        Ok(ok(format!("skill updated: {}", s.id), true))
                    }
                    "delete" => {
                        let (skills_list, _) = skills::list_skills(paths);
                        let hit = pos.get(1).and_then(|q| {
                            skills_list.iter().find(|s| &s.id == q || &s.name == q)
                        });
                        let Some(hit) = hit else {
                            return Ok(bad(format!(
                                "no skill {}",
                                pos.get(1).cloned().unwrap_or_else(|| "(missing id)".into())
                            )));
                        };
                        skills::delete_skill(paths, &hit.id)?;
                        Ok(ok(format!("skill deleted: {}", hit.id), true))
                    }
                    _ => Ok(bad("skill: expected show|create|edit|delete".into())),
                }
            };
            match run() {
                Ok(r) => r,
                Err(e) => bad(e),
            }
        }

        _ => bad(format!("unknown command {} — try `help`", js_json_str(&cmd))),
    }
}
